### PR TITLE
engine: implement execution engine

### LINKED ISSUE #177: engine: implement engine
Implement an engine that can execute our IR.

Scope:
- [ ] define file format (#178)
- [ ] implement support for file format (#180)
- [x] add high level profiling (#181)
- [x] add integration tests (#182)
- [ ] add benchmarks (#183)
- [ ] add documentation documents

### PR DESCRIPTION
**Pull request by TimSatke**
Implement the engine (former executor) to take in an IR and represent the stated changes in a file format.

### CODE REVIEW

File: internal/engine/error.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	ErrClosed Error = "already closed"
+	// ErrUnsupported indicates that a requested feature is explicitely not
+	// supported. This is different from ErrUnimplemented, since
+	// ErrUnimplemented indicates, that the feature has not been implemented
+	// yet, while ErrUnsupported indicates, that the feature is intentionally
+	// unimplemented.
+	ErrUnsupported Error = "unsupported"
+)
+
+// ErrNoSuchFunction returns an error indicating that an error with the given
```
Comment (TimSatke) [inline internal/engine/error.go:26]:
```suggestion
// ErrNoSuchFunction returns an error indicating that a function with the given
```

File: internal/engine/evaluate.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		}
+		var colName string
+		if colNameExpr.Is(types.String) {
+			colName = colNameExpr.(types.StringValue).Value
+		}
+		if col.Table != "" {
+			colName = col.Table + "." + colName
+		}
+
+		// TODO: support alias
```
Comment (TimSatke) [inline internal/engine/evaluate.go:67]: Create follow-up ticket

File: internal/engine/expression.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	"strconv"
+
+	"github.com/tomarrell/lbadd/internal/compiler/command"
+	"github.com/tomarrell/lbadd/internal/engine/types"
+)
+
+// evaluateExpression evaluates the given expression to a runtime-constant
+// value, meaning that it can only be evaluated to a constant value with a given
+// execution context. This execution context must be inferred from the engine
+// receiver.
```
Comment (TimSatke) [inline internal/engine/expression.go:14]:
```suggestion
// execution context.
```

File: internal/engine/expression.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+func (e Engine) evaluateExpression(ctx ExecutionContext, expr command.Expr) (types.Value, error) {
+	switch ex := expr.(type) {
+	case command.ConstantBooleanExpr:
+		return types.NewBool(ex.Value), nil
+	case command.LiteralExpr:
+		return e.evaluateLiteralExpr(ctx, ex)
+	case command.FunctionExpr:
+		return e.evaluateFunctionExpr(ctx, ex)
+	}
+	return nil, fmt.Errorf("cannot evaluate expression of type %T", expr)
```
Comment (TimSatke) [inline internal/engine/expression.go:24]:
```suggestion
	return nil, ErrUnimplemented(fmt.Sprintf("evaluate %T", expr))
```

File: internal/engine/numeric_parser.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	isErronous              bool
+	hasDigitsBeforeExponent bool
+
+	value *bytes.Buffer
+
+	current numericParserState
+}
+
+// ToNumericValue checks whether the given string is of this form
+// https://www.sqlite.org/lang_expr.html#literal_values_constants_ . If it is, a
```
Comment (TimSatke) [inline internal/engine/numeric_parser.go:28]:
```suggestion
// https://www.sqlite.org/lang_expr.html#literal_values_constants_ . If it is, an
```

File: internal/engine/scan.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+func (e Engine) scanSimpleTable(ctx ExecutionContext, table command.SimpleTable) (Table, error) {
+	tableName := table.QualifiedName()
+
+	// only perform scan if not already scanned
+	if table, alreadyScanned := ctx.getScannedTable(tableName); alreadyScanned {
+		return table, nil
+	}
+
+	// TODO: load table from the database file
```
Comment (TimSatke) [inline internal/engine/scan.go:13]: Create follow-up ticket

File: internal/engine/storage/page_manager.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("load: %w", err)
+	}
+	return p, nil
+}
+
+// WritePage will write the given page into secondary storage. It is guaranteed,
+// that after this call returns, the page is present on disk.
+func (m *PageManager) WritePage(p *page.Page) error {
+	data := p.RawData() // TODO: avoid copying in RawData()
```
Comment (TimSatke) [inline internal/engine/storage/page_manager.go:53]: Remove TODO

File: internal/engine/storage/db_test.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+	// load the header page, which is the first page (offset 0) in the file
+	headerPage := loadPageFromOffset(assert, f, 0)
+	assert.EqualValues(3, headerPage.CellCount())
+	assert.Equal(
+		encodeUint64(3),
+		mustCell(assert, headerPage, HeaderPageCount).(page.RecordCell).Record,
+	)
+
+	// Allocating a new page must persist it in the created database file. This
```
Comment (SUMUKHA-PK) [inline internal/engine/storage/db_test.go:36]:
```suggestion
	// Allocating a new page must persist it is in the created database file. This
```

**Pull request merged**